plugins/serializers/binary/entry.go

```diff
@@ -0,0 +1,296 @@
+	ReadFrom         string `toml:"read_from"`         // field, tag, time, name
+	Name             string `toml:"name"`              // name of entry
+	DataFormat       string `toml:"data_format"`       // int8, int16, int32, int64, uint8, uint16, uint32, uint64, float32, float64, string
+	StringTerminator string `toml:"string_terminator"` // for string metrics: null, 0x00, 00, ....
+	StringLength     uint64 `toml:"string_length"`     // for string only, target size
+	TimeFormat       string `toml:"time_format"`       // for time metrics: unix, unix_ms, unix_us, unix_ns
+
+	bytes       uint64
+	termination byte
+}
+
+func (e *Entry) FillDefaults() error {
+	// Normalize
+	e.ReadFrom = strings.ToLower(e.ReadFrom)
+
+	if e.ReadFrom == "" {
+		e.ReadFrom = "field"
+	}
+
+	switch e.ReadFrom {
+	case "field", "tag":
+		if e.Name == "" {
+			return errors.New("missing name")
+		}
+	case "time":
+		switch e.TimeFormat {
+		// 'unix' the default
+		case "":
+			e.TimeFormat = "unix"
+		// Plugin specific names
+		case "unix", "unix_ms", "unix_us", "unix_ns":
+		default:
+			return errors.New("invalid time format")
+		}
+	case "name":
+	default:
+		return fmt.Errorf("unknown assignment %q", e.ReadFrom)
+	}
+
+	// Check data format
+	switch e.DataFormat {
+	case "":
+		return errors.New("missing data format")
+	case "int64", "uint64", "float64":
+		e.bytes = 8
+	case "int32", "uint32", "float32":
+		e.bytes = 4
+	case "int16", "uint16":
+		e.bytes = 2
+	case "int8", "uint8":
+		e.bytes = 1
+	case "string":
+		if e.StringLength < 1 {
+			return errors.New("string length must be at least 1")
+		}
+
+		e.bytes = e.StringLength
+	}
+
+	// Check string terminator
+	switch e.StringTerminator {
+	case "", "null":
+		e.termination = 0x00
+	default:
+		terminatorLength := len(e.StringTerminator)
+
+		// support both 0xXX and XX
+		termination, err := hex.DecodeString(e.StringTerminator[terminatorLength-2:])
+
+		if err != nil {
+			return fmt.Errorf("decoding terminator failed for %q: %w", e.Name, err)
+		}
+
+		if len(termination) != 1 {
+			return fmt.Errorf("terminator must be a single byte, got %q", e.StringTerminator)
+		}
+
+		e.termination = termination[0]
+	}
+
+	return nil
+}
+
+func (e *Entry) timeToTimestamp(t time.Time) int64 {
+	var timestamp int64
+
+	switch e.TimeFormat {
+	case "unix":
+		timestamp = t.Unix()
+	case "unix_ms":
+		timestamp = t.UnixMilli()
+	case "unix_us":
+		timestamp = t.UnixMicro()
+	case "unix_ns":
+		timestamp = t.UnixNano()
+	}
+
+	return timestamp
+}
+
+func appendToSlice[T any](arr []T, targetLength uint64, value T) []T {
+	for uint64(len(arr)) < targetLength {
+		arr = append(arr, value)
+	}
+
+	return arr
+}
+
+func (e *Entry) SerializeValue(value interface{}, converter binary.ByteOrder) ([]byte, error) {
+	if e.DataFormat != "string" {
+		// time to int64
+		switch v := value.(type) {
+		case time.Time:
+			value = e.timeToTimestamp(v)
+		case string:
+			var err error
+			value, err = strconv.ParseFloat(v, 64)
```

Comment:
A bit weird to have this string to float conversion when the value is for example "1234678", but then if DataFormat is `int` we get a warning about likely data loss..

Reply:
@Hipska  Would something like this (moved to helper method) be any better?

```go
if e.DataFormat != "string" {
  // time to int64
  switch v := value.(type) {
  case time.Time:
	  value = e.timeToTimestamp(v)
  case string:
	  var err error
  
	  if strings.Contains(v, ".") {
		  value, err = strconv.ParseFloat(v, 64)
	  } else if strings.Contains(v, "-") {
		  value, err = strconv.ParseInt(v, 0, 64)
	  } else {
		  value, err = strconv.ParseUint(v, 0, 64)
	  }
  
	  if err != nil {
		  return nil, fmt.Errorf("string to float conversion failed: %w", err)
	  }
  }
}
```

Reply:
Yeah, that will not give dataloss warning in most cases..